File: solutions/java/src/main/java/com/epi/MultibetCardColorGame.java
package com.epi;

public class MultibetCardColorGame {
  // @include
  public static double computeBestPayoff(int cash) {
    double upperBound = 9.09 * cash;
    int numRed = 26;
    int numCards = 52;
    Double[][][] cache = new Double[(int)upperBound][numRed + 1][numCards + 1];
    return computeBestPayoffHelper(cache, upperBound, cash, 26, 52);
  }

  private static double computeBestPayoffHelper(Double[][][] cache,
                                                double upperBound, int cash,
                                                int numRed, int numCards) {
    if (cash >= upperBound) {
      return cash;
    }

    if (numRed == numCards || numRed == 0) {
      return cash * Math.pow(2, numCards);
    }

    if (cache[cash][numRed][numCards] == null) {
      double best = Double.MIN_VALUE;
      for (int bet = 0; bet <= cash; ++bet) {
        double redLowerBound
            = Math.min(computeBestPayoffHelper(cache, upperBound, cash + bet,
                                               numRed - 1, numCards - 1),
                       computeBestPayoffHelper(cache, upperBound, cash - bet,
                                               numRed, numCards - 1));

        double blackLowerBound
            = Math.min(computeBestPayoffHelper(cache, upperBound, cash - bet,
                                               numRed - 1, numCards - 1),
                       computeBestPayoffHelper(cache, upperBound, cash + bet,
                                               numRed, numCards - 1));
        best = Math.max(best, Math.max(redLowerBound, blackLowerBound));
      }

      cache[cash][numRed][numCards] = best;
    }

    return cache[cash][numRed][numCards];
  }
  // @exclude

  public static void main(String[] args) {
    int ans = (int)computeBestPayoff(100);
    assert ans == 808;
    System.out.println("100 cash can get " + ans);
  }
}

File: solutions/cpp/Multibet_card_color_game.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>
#include <unordered_map>

using std::cout;
using std::endl;
using std::max;
using std::min;
using std::numeric_limits;
using std::unordered_map;

double compute_best_payoff_helper(
    unordered_map<int, unordered_map<int, unordered_map<int, double>>>& cache,
    double upper_bound, int cash, int num_red, int num_cards);

// @include
double compute_best_payoff(int cash) {
  double upper_bound = 9.09 * cash;
  unordered_map<int, unordered_map<int, unordered_map<int, double>>> cache;
  return compute_best_payoff_helper(cache, upper_bound, cash, 26, 52);
}

double compute_best_payoff_helper(
    unordered_map<int, unordered_map<int, unordered_map<int, double>>>& cache,
    double upper_bound, int cash, int num_red, int num_cards) {
  if (cash >= upper_bound) {
    return cash;
  }

  if (num_red == num_cards || num_red == 0) {
    return cash * pow(2, num_cards);
  }

  if (cache[cash][num_red].find(num_cards) == cache[cash][num_red].end()) {
    double best = numeric_limits<double>::min();
    for (int bet = 0; bet <= cash; ++bet) {
      double red_lower_bound =
          min(compute_best_payoff_helper(cache, upper_bound, cash + bet,
                                         num_red - 1, num_cards - 1),
              compute_best_payoff_helper(cache, upper_bound, cash - bet,
                                         num_red, num_cards - 1));

      double black_lower_bound =
          min(compute_best_payoff_helper(cache, upper_bound, cash - bet,
                                         num_red - 1, num_cards - 1),
              compute_best_payoff_helper(cache, upper_bound, cash + bet,
                                         num_red, num_cards - 1));
      best = max({best, red_lower_bound, black_lower_bound});
    }
    cache[cash][num_red][num_cards] = best;
  }
  return cache[cash][num_red][num_cards];
}
// @exclude

int main(int argc, char* argv[]) {
  int ans = compute_best_payoff(100);
  assert(ans == 808);
  cout << "100 cash can get " << ans << endl;
  return 0;
}
